test/uk/gov/hmrc/agentservicesaccount/controllers/AgentServicesControllerSpec.scala

```diff
@@ -16,66 +16,160 @@
 
   "root" should {
     "return Status: OK and body containing correct content" in {
-      val arn = "TARN0000001"
-      val authActions = new AuthActions(null, null, null) {
-        override def AuthorisedWithAgentAsync(body: AsyncPlayUserRequest): Action[AnyContent] =
-          Action.async { implicit request =>
-            body(AgentRequest(Arn(arn), request))
-          }
-      }
 
-      val controller = new AgentServicesController(messagesApi, authActions, externalUrls)
+
+      val controller = new AgentServicesController(messagesApi, authActions, continueUrlActions, externalUrls)
 
       val response = controller.root()(FakeRequest("GET", "/"))
 
       status(response) shouldBe OK
       contentType(response).get shouldBe HTML
       val content = contentAsString(response)
       content should include("Agent Services account")
-      content should not include("Agent Services Account")
+      content should not include "Agent Services Account"
       content should include(messagesApi("agent.services.account.heading"))
       content should include(messagesApi("agent.services.account.heading.summary"))
       content should include(arn)
       content should include(signOutUrl)
     }
 
     "return the redirect returned by authActions when authActions denies access" in {
-      val authActions = new AuthActions(null, null, null) with Results {
-        override def AuthorisedWithAgentAsync(body: AsyncPlayUserRequest): Action[AnyContent] =
-          Action { implicit request =>
-            Redirect("/gg/sign-in", 303)
+
+      val authActions = new AuthActions(null, null, null) {
+        override def AuthorisedWithAgentAsync = new ActionBuilder[AgentRequest] {
+          override def invokeBlock[A](request: Request[A], block: (AgentRequest[A]) => Future[Result]): Future[Result] = {
+            Future.successful(Redirect("/gg/sign-in", 303))
           }
+        }
       }
 
-      val controller = new AgentServicesController(messagesApi, authActions, externalUrls)
+      val controller = new AgentServicesController(messagesApi, authActions, continueUrlActions, externalUrls)
 
       val response = controller.root()(FakeRequest("GET", "/"))
 
       status(response) shouldBe 303
       redirectLocation(response) shouldBe Some("/gg/sign-in")
     }
+
+    "do not fail without continue url parameter" in {
+      val controller = new AgentServicesController(messagesApi, authActions, continueUrlActions, externalUrls)
+      val response = controller.root().apply(FakeRequest("GET", "/"))
+      status(response) shouldBe OK
+      contentType(response).get should not include "<button class=\"btn button\" type=\"continue\""
+    }
+
+    "support relative continue url parameter" in {
+      val controller = new AgentServicesController(messagesApi, authActions, continueUrlActions, externalUrls)
+      val response = controller.root().apply(FakeRequest("GET", "/?continue=/foo"))
```

Comment:
we use url encoding only for test urls having query part, `/` doesn't need encoding